adafruit_connection_manager.py

```diff
@@ -0,0 +1,402 @@
+        try:
+            return next(
+                key for key, value in self._open_sockets.items() if value == socket
+            )
+        except StopIteration:
+            return None
+
+    def close_socket(self, socket: SocketType) -> None:
+        """Close a previously opened socket."""
+        logger.debug("ConnectionManager.close_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  closing {key}")
+        socket.close()
+        del self._socket_free[socket]
+        del self._open_sockets[key]
+
+    def free_socket(self, socket: SocketType) -> None:
+        """Mark a previously opened socket as free so it can be reused if needed."""
+        logger.debug("ConnectionManager.free_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  flagging {key} as free")
+        self._socket_free[socket] = True
+
+    # pylint: disable=too-many-locals,too-many-statements
+    def get_socket(
+        self,
+        host: str,
+        port: int,
+        proto: str,
+        session_id: Optional[str] = None,
+        *,
+        timeout: float = 1,
+        is_ssl: bool = False,
+        ssl_context: Optional[SSLContextType] = None,
+        max_retries: int = 5,
+        exception_passthrough: bool = False,
+    ) -> CircuitPythonSocketType:
+        """Get a new socket and connect"""
+        # pylint: disable=too-many-branches
+        logger.debug("ConnectionManager.get_socket()")
+        logger.debug(f"  tracking {len(self._open_sockets)} sockets")
+        key = (host, port, proto, str(session_id))
+        logger.debug(f"  getting socket for {key}")
+        if key in self._open_sockets:
+            socket = self._open_sockets[key]
+            if self._socket_free[socket]:
+                logger.debug(f"  found existing {key}")
+                self._socket_free[socket] = False
+                return socket
+
+        if proto == "https:":
+            is_ssl = True
+        if is_ssl and not ssl_context:
+            raise RuntimeError(
+                "ssl_context must be set before using adafruit_requests for https"
+            )
+
+        addr_info = self._socket_pool.getaddrinfo(
+            host, port, 0, self._socket_pool.SOCK_STREAM
+        )[0]
+
+        retry_count = 0
+        socket = None
+        last_exc = None
+        last_exc_new_type = None
+        while retry_count < max_retries and socket is None:
+            if retry_count > 0:
+                logger.debug(f"  retry #{retry_count}")
+                if any(self._socket_free.items()):
+                    self._free_sockets()
+                else:
+                    raise RuntimeError("Sending request failed") from last_exc
+            retry_count += 1
```

Comment:
@tannewt I have a question on how you would like to fix this. In theory we try 5 times and if we fail we raise with `Repeated socket failures`, but that will never happen.

If it's the first request:
Try one: error
Try two: `any(self._socket_free.items())` returns false, error right away

If it's not the first request:
Try one: error
Try two: `any(self._socket_free.items())` returns true, clears `_socket_free`, error
Try three: `any(self._socket_free.items())` returns false, error right away

The only case it would try 5 times is if you have an open and active socket, which seems an odd case to keep trying. Thoughts? Leave it as is for now?

Reply:
Am writing tests and this made sense to change to:
```
if any(socket for socket, free in self._socket_free.items() if free is True)
```
to avoid extra loops but then couldn't get that last error at all

Reply:
Sorry, my brain doesn't have enough context to be concrete. It seems like the retrying here should be robust enough that minimqtt doesn't need to do any retrying itself.

Reply:
Okay, so 2 separate issues.

1. This code was pulled from `requests` and has built in retry, that's flawed
2. minimqtt also built in a retry mechanism that is similar but different on what they feel is retry-able

It might be easiest to look at the minimqtt PR listed above to see what I mean.

I'm happy to work towards finding something that works for all libraries, but don't know how to get  consensus